### PR TITLE
MAINT: Prepare cdist/pdist for C++ rework

### PR DESCRIPTION
**Pull request by peterbell10**
#### Reference issue
Related to https://github.com/scipy/scipy/issues/13629#issuecomment-790831819

#### What does this implement/fix?
In preparation for rewriting the C++ implementations for `cdist`/`pdist`, I have taken the logic that's closely tied to the current C implementations and moved it from `cdist`/`pdist` into the per-metric `MetricInfo` object. This means I can incrementally update `cdist`/`pdist` one metric at a time in future PRs. 

#### Additional information

Benchmark changes (including weighted benchmarks in #13785):
```python
       before           after         ratio
     [7dcdeffe]       [4769878d]
     <master>         <xdist-python-rework>
+     4.68±0.02μs      5.15±0.05μs     1.10  spatial.Xdist.time_cdist(10, 'chebyshev')
+     2.55±0.01ms      2.80±0.01ms     1.10  spatial.XdistWeighted.time_pdist(20, 'jaccard')
+     5.22±0.04μs       5.62±0.1μs     1.08  spatial.Xdist.time_cdist(10, 'rogerstanimoto')
+     4.62±0.06μs      4.97±0.03μs     1.08  spatial.Xdist.time_cdist(10, 'sqeuclidean')
+     4.83±0.06μs      5.16±0.01μs     1.07  spatial.Xdist.time_cdist(10, 'cosine')
+      5.26±0.1μs      5.62±0.04μs     1.07  spatial.Xdist.time_cdist(10, 'kulsinski')
+     5.34±0.06μs      5.67±0.02μs     1.06  spatial.Xdist.time_cdist(10, 'yule')
+     4.96±0.04μs      5.23±0.03μs     1.05  spatial.Xdist.time_cdist(10, 'jaccard')
-      11.4±0.1μs      10.8±0.05μs     0.95  spatial.Xdist.time_pdist(100, 'sqeuclidean')
-      11.4±0.3μs       10.8±0.1μs     0.95  spatial.Xdist.time_pdist(100, 'cityblock')
-     4.66±0.02ms      4.41±0.01ms     0.95  spatial.XdistWeighted.time_pdist(20, 'dice')
-      3.83±0.1ms      3.60±0.01ms     0.94  spatial.XdistWeighted.time_pdist(20, 'canberra')
-     4.48±0.06μs       4.14±0.1μs     0.92  spatial.Xdist.time_pdist(10, 'jaccard')
-     5.94±0.09μs      5.46±0.01μs     0.92  spatial.Xdist.time_pdist(10, 'jensenshannon')
-      40.0±0.4μs       36.8±0.6μs     0.92  spatial.Xdist.time_cdist(100, 'hamming')
-      21.7±0.7μs      19.9±0.07μs     0.92  spatial.Xdist.time_pdist(10, 'seuclidean')
-      4.45±0.1μs      4.02±0.03μs     0.91  spatial.Xdist.time_pdist(10, 'chebyshev')
-     4.47±0.01μs      4.04±0.04μs     0.90  spatial.Xdist.time_pdist(10, 'cityblock')
-     4.60±0.07μs      4.13±0.05μs     0.90  spatial.Xdist.time_pdist(10, 'russellrao')
-        16.5±1μs       14.8±0.9μs     0.90  spatial.XdistWeighted.time_cdist(20, 'hamming')
-      4.73±0.3μs      4.22±0.07μs     0.89  spatial.Xdist.time_pdist(10, 'kulsinski')
-     4.48±0.01μs      4.00±0.01μs     0.89  spatial.Xdist.time_pdist(10, 'euclidean')
-     4.65±0.03μs      4.12±0.06μs     0.89  spatial.Xdist.time_pdist(10, 'sokalsneath')
-     4.49±0.06μs      3.97±0.01μs     0.89  spatial.Xdist.time_pdist(10, 'canberra')
-      4.48±0.2μs      3.96±0.04μs     0.88  spatial.Xdist.time_pdist(10, 'sqeuclidean')
-      4.86±0.6μs      4.25±0.02μs     0.87  spatial.Xdist.time_pdist(10, 'dice')
-     18.4±0.06μs      15.8±0.07μs     0.86  spatial.XdistWeighted.time_cdist(10, 'minkowski')
-      19.1±0.9μs       16.3±0.1μs     0.85  spatial.XdistWeighted.time_cdist(10, 'minkowski-P3')
-      15.1±0.4μs       12.8±0.5μs     0.85  spatial.Xdist.time_cdist(10, 'hamming')
-      12.3±0.2μs      10.3±0.06μs     0.83  spatial.Xdist.time_cdist(10, 'minkowski-P3')
-      5.10±0.5μs       4.20±0.1μs     0.82  spatial.Xdist.time_pdist(10, 'yule')
-      15.6±0.2μs      12.5±0.04μs     0.80  spatial.XdistWeighted.time_cdist(10, 'hamming')
-     7.53±0.07μs       5.49±0.2μs     0.73  spatial.Xdist.time_cdist(10, 'minkowski')
-         504±1μs       10.5±0.2μs     0.02  spatial.XdistWeighted.time_pdist(10, 'chebyshev')
-     1.09±0.05ms       11.8±0.5μs     0.01  spatial.XdistWeighted.time_cdist(10, 'chebyshev')
-     2.04±0.02ms      11.0±0.09μs     0.01  spatial.XdistWeighted.time_pdist(20, 'chebyshev')
-     4.31±0.06ms       12.6±0.2μs     0.00  spatial.XdistWeighted.time_cdist(20, 'chebyshev')
-      53.1±0.2ms       23.6±0.6μs     0.00  spatial.XdistWeighted.time_pdist(100, 'chebyshev')
-       107±0.4ms       39.0±0.8μs     0.00  spatial.XdistWeighted.time_cdist(100, 'chebyshev')
```

Summary: Overhead for small sizes has changed +/- .5 us for some metrics, but more improved than regressed. Weighted chebyshev also improved massively because the weighted kernels existed but weren't actually called before.

### CODE REVIEW

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -2653,57 +2889,30 @@ def cdist(XA, XB, metric='euclidean', **kwargs):
...
-        XA, XB, typ, kwargs = _validate_cdist_input(XA, XB, mA, mB, n,
-                                                    metric_name, **kwargs)
-
-        for i in range(0, mA):
-            for j in range(0, mB):
-                dm[i, j] = metric(XA[i], XB[j], **kwargs)
-
+        metric_info = _METRIC_ALIAS.get(mstr, None)
+        if metric_info is not None:
+            XA, XB, typ, kwargs = _validate_cdist_input(
```
Comment (tupui) [inline scipy/spatial/distance.py:2897]: The `metric_info is not None` is already done in `_validate_cdist_input ` no?

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -1959,69 +2189,42 @@ def pdist(X, metric='euclidean', **kwargs):
...
-                                                   metric_name, **kwargs)
+        metric_info = _METRIC_ALIAS.get(mstr, None)
 
-        k = 0
-        for i in range(0, m - 1):
-            for j in range(i + 1, m):
-                dm[k] = metric(X[i], X[j], **kwargs)
-                k = k + 1
+        if metric_info is not None:
+            X, typ, kwargs = _validate_pdist_input(
```
Comment (tupui) [inline scipy/spatial/distance.py:2204]: Same here.

**Review by tupui — COMMENTED** [review]
Good refactoring!

Maybe we should separate the actual metrics present in this file from the infrastructure part?

To address further performances, what do you think about making the validation part optional? By calling a private `_cdist` which would be striped of validations for instance. This way, if someone needs to use this in a loop it would help.

**Comment by peterbell10** [comment]
>Maybe we should separate the actual metrics present in this file from the infrastructure part?

I plan on moving them into C++ anyway, so IMO not worth the churn of moving now.

>To address further performances, what do you think about making the validation part optional? By calling a private _cdist which would be striped of validations for instance. This way, if someone needs to use this in a loop it would help.

So we're on the same page, input validation isn't causing the slowdown in #13629. It's falling back to the `test_` versions because weighted metrics aren't implemented in C++ that makes it slower.

A few other observations:
1. For most metrics, the validation is fairly trivial e.g. just checking shapes are compatible. There's very little to be gained there.
2. For some metrics the "validator" actually forms a key part of the calculation, e.g. `seuclidean` and `mahalanobis` use it to calculate the default `V` and `VI` values. Without which, the code doesn't work.
3. `_validate_weights` may have a more significant impact since it reads the actual values to ensure weights are non-negative. However, it's the wrong place to look for optimization  when weighted metrics are already 100x slower than they could be for other reasons.

In summary, I think it's the wrong place to be focusing right now.

File: scipy/spatial/distance.py (hunk 1 above)
Comment (peterbell10) [reply scipy/spatial/distance.py:2895]: Removed the check from `_validate_{c,p}dist_input`. All other call sites know for a fact the `metric_info` is valid, which makes more sense to me.

**Review by tupui — APPROVED** [review]
Thanks for the detailed answer!

For the validation part, I did some quick tests in #13761 and `np.asarray` is responsible for up to 10-20% of the computation time of a cythonized function.

But I agree when you say that here there are more pressing things. Since we are refactoring, I just wanted to highlight this and maybe include something along the line.

I am approving this as there will be a some follow up work (and this should not block your progress). This already improves performance and clean the code which is great!

File: scipy/spatial/distance.py [hunk 3]
```diff
@@ -2350,7 +2547,42 @@ def num_obs_y(Y):
...
+    return dm
+
+
+def _cdist_callable(XA, XB, *, out, metric, **kwargs):
+    mA = XA.shape[0]
+    mB = XB.shape[0]
+    dm = _prepare_out_argument(out, np.double, (mA, mB))
+    for i in range(mA):
+        for j in range(mB):
+            dm[i, j] = metric(XA[i], XB[j], **kwargs)
```
Comment (tylerjereddy) [inline scipy/spatial/distance.py:2581]: I hadn't realized there was pure Python management of the iteration like this--are you planning to move that out of Python as well? Anyway, that hasn't changed before/after here.

**Review by tylerjereddy — APPROVED** [review]
Ok, CI is passing except for an unrelated timeout and there's another approving review.

Admittedly, the diff is quite large so I'm leaning pretty heavily on the CI/testing and the fact that no tests have been changed at all here.

I did try to find the small % of lines that codecov claims are not covered in the diff, but it was unfortunately nonsense when I looked at the uncovered lines in this case.

**Comment by tylerjereddy** [comment]
Thanks @peterbell10 and @tupui

File: scipy/spatial/distance.py (hunk 3 above)
Comment (tylerjereddy) [reply scipy/spatial/distance.py:2581]: Is this as bad as it looks? Every row in each input of points has to be traversed in raw Python then calling into the C function for all those pairs? Or only happening for some metrics/situations?

Comment (peterbell10) [reply scipy/spatial/distance.py:2581]: This is meant to allow user-provided metric-functions. In which case I don't think there's any simple way to speed it up. Really it's the API that would need to change for that to work.
